rasa/nlu/featurizers/dense_featurizer/lm_featurizer.py

```diff
@@ -754,13 +796,19 @@ def train(
                     self._set_lm_features(batch_docs[index], ex, attribute)
                 batch_start_index += batch_size
 
-    def process(self, message: Message, **kwargs: Any) -> None:
-        """Process an incoming message by computing its tokens and dense features.
+    def process(self, messages: List[Message]) -> None:
```

Comment:
```suggestion
    def process(self, messages: List[Message]) -> List[Message]:
```

Reply:
you have to return the messages at the end of it

Reply:
same in `process_training_data`

Reply:
is it actually possible to unify `process` and `process_training_data`? 🤔

Reply:
Yeah we could (I was also wondering if I should do that for the ConveRT featurizer but because of all the other stuff I had changed, I left it as is eventually 😅 ). 

But it doesn't matter at the moment - if we use this either with 1 or 0 messages only.

And I still don't really understand (or keep forgetting) why it is 0 or 1 message. If there were 0 new messages, then we should get the same output from the pipeline... ? - Does it have to do with the special restart events? But even those extract those things from the message 🤔

Reply:
It's currently passing by reference and modifying messages in place. Should modify and return a copy instead? 🤔

Reply:
>is it actually possible to unify process and process_training_data? 🤔

Need to go through the code to understand what's it's doing. 
@ka-bu is your response to tobi about this too?

Reply:
If it's too much effort unifying this then let's skip it.

> you have to return the messages at the end of it
It's currently passing by reference and modifying messages in place, isn't it? Should modify and return a copy instead? 🤔

Modifying in-place is good for now. But we still need to return the things as output to work with the graph (every node needs to have input and output). So no copy is needed

> If there were 0 new messages, then we should get the same output from the pipeline... ? - Does it have to do with the special restart events? But even those extract those things from the message 🤔

We have the 0 message case when you predict an action after having just predicted an action, e.g:
```
- User: "Hi"
- action: utter_greet  <-- in this case we have list with message [1]
- action: action_listen  <-- in this case we predict with an empty list of messages. All NLU components just skip in this case
```

Reply:
> We have the 0 message case when you predict an action after having just predicted an action, e.g:

- If the policies do not see that new action (- action: utter_greet) then that means they only see (- User: "Hi") and they will predict the exact same action (- action: utter_greet) again and again. 
- If the policies see that new action (- action: utter_greet), then that means a message containing ACTION_NAME utter_greet has arrived, no?
Or do you mean the prediction *after* (- action: action_listen) - but that doesn't need to be passed through the graph anyway, no?

Reply:
@aciel - the response was about replacing how process works (i.e. process message by message) by how process_training_data works (i.e. processing a bunch of messages in parallel at a time - at most 64 to be precise ^^). It's not much effort unifying it but it's pointless if process is only called with 1 or 0 messages at a time.